### PR TITLE
br: support restore snapshot and log(PiTR) only by one-command

### PR DESCRIPTION
**Pull request by joccau**
Signed-off-by: joccau <zak.zhao@pingcap.com>

### What problem does this PR solve?

Issue Number: close #xxx

### What is changed and how it works?
- Support restore snapshot and log(PiTR) only by one-command, 
   It will restore snapshot firstly, restore log at the time range [snapshot-ts, restored-ts].

`
    restore point --full-backup-storage "snapshot-storage" 
    --storage "log-storage" --restored-ts=400036290571534337
`
- If it has a gap between snapshot-timestamp of `backup full` and start-timestamp of `backup log`, report an error and a selective plan:

 `br restore full -s "snapshot-storage"`
`br restore point --start-ts  --restored-ts  -s "log-storage" `

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/task/restore.go [hunk 1]
```diff
@@ -317,10 +334,10 @@ func IsStreamRestore(cmdName string) bool {
 }
 
 // RunRestore starts a restore task inside the current goroutine.
-func RunRestore(c context.Context, g glue.Glue, cmdName string, cfg *RestoreConfig) error {
+func RunRestore(c context.Context, g glue.Glue, cmdName string, cfg *RestoreConfig) (uint64, error) {
```
Comment (3pointer) [inline br/pkg/task/restore.go:337]: I think you can add a defer function
if err != nil then return 0, else return endversion.

File: br/pkg/restore/client.go [hunk 2]
```diff
@@ -1924,3 +1928,21 @@ func (rc *Client) UpdateSchemaVersion(ctx context.Context) error {
 
 	return nil
 }
+
+func CalculateGlobalCheckpointTs(ms []*backuppb.Metadata) uint64 {
```
Comment (3pointer) [inline br/pkg/restore/client.go:1932]: This function seems unused?

File: br/pkg/task/restore.go [hunk 3]
```diff
@@ -167,16 +173,27 @@ func DefineStreamRestoreFlags(command *cobra.Command) {
...
 	}
 	if cfg.RestoreTS, err = ParseTSString(tsString); err != nil {
 		return errors.Trace(err)
 	}
+
 	if cfg.FullBackupStorage, err = flags.GetString(FlagStreamFullBackupStorage); err != nil {
 		return errors.Trace(err)
 	}
+	if len(cfg.FullBackupStorage) == 0 {
+		return errors.Annotatef(err, "failed to get flag %s", FlagStreamFullBackupStorage)
```
Comment (3pointer) [inline br/pkg/task/restore.go:195]: don't we support split restore full and restore log?

File: br/pkg/restore/client.go [hunk 4]
```diff
@@ -1816,7 +1814,13 @@ func (rc *Client) RestoreMetaKVFile(
...
 		if err != nil {
 			return errors.Trace(err)
-		} else if ts > rc.restoreTs {
+		}
+
+		// The commitTs in write CF need be limited on [startTs, restoreTs].
+		// We can restore more key-value in default CF.
+		if ts > rc.restoreTs {
+			continue
+		} else if file.Cf == "write" && ts < rc.startTS {
```
Comment (YuJuncen) [inline br/pkg/restore/client.go]: merge them to the same `if`?

File: br/pkg/task/restore.go [hunk 5]
```diff
@@ -45,7 +45,8 @@ const (
 	// current only support STRICT or IGNORE, the default is STRICT according to tidb.
 	FlagWithPlacementPolicy = "with-tidb-placement-mode"
 
-	// FlagStreamRestoreTS is used for log restore.
+	// FlagStreamStartTS and FlagStreamRestoreTS is used for log restore timestamp range.
+	FlagStreamStartTS   = "start-ts"
```
Comment (YuJuncen) [inline br/pkg/task/restore.go:49]: Should that be called as `snapshot-ts`?

File: br/pkg/task/stream.go [hunk 6]
```diff
@@ -892,17 +897,48 @@ func RunStreamRestore(
...
+				cfg.StartTS, oracle.GetTimeFromTS(cfg.StartTS),
+				logStartTS, oracle.GetTimeFromTS(logStartTS))
+		}
 	}
 
-	cfg.Config.Storage = StreamStorage
-	if err := restoreStream(ctx, g, cmdName, cfg); err != nil {
+	if logStartTS < truncateTS {
+		logStartTS = truncateTS
+	}
```
Comment (YuJuncen) [inline br/pkg/task/stream.go:921]: should we do that before we check the relationship between `cfg.StartTS` and `logStartTS`?

File: br/pkg/task/stream.go [hunk 7]
```diff
@@ -927,73 +962,21 @@ func restoreStream(
...
-
-	if err = client.SetStorage(ctx, u, &opts); err != nil {
-		return errors.Trace(err)
-	}
-	client.SetRateLimit(cfg.RateLimit)
-	client.SetCrypter(&cfg.CipherInfo)
-	client.SetConcurrency(uint(cfg.Concurrency))
-	client.SetSwitchModeInterval(cfg.SwitchModeInterval)
-	safePoint := client.GetTruncateSafepoint(ctx)
-	if cfg.RestoreTS < safePoint {
```
Comment (3pointer) [inline br/pkg/task/stream.go:966]: this logic has been removed?

**Review by 3pointer — APPROVED** [review]
LGTM

**Review by YuJuncen — APPROVED** [review]
rest lgtm

**Pull request merged**